domain/application/state/application.go

```diff
@@ -1874,3 +1874,78 @@ func decodeArchitecture(arch int) (application.Architecture, error) {
 		return -1, fmt.Errorf("unsupported architecture: %d", arch)
 	}
 }
+
+// GetApplicationIDByUnitName returns the application ID for the named unit,
+// returning an error satisfying [applicationerrors.UnitNotFound] if the unit
+// doesn't exist.
+func (st *ApplicationState) GetApplicationIDByUnitName(
+	ctx context.Context,
+	name coreunit.Name,
+) (coreapplication.ID, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	queryUnit := `
+SELECT application_uuid AS &applicationID.uuid
+FROM unit
+WHERE name = $unitNameAndUUID.name
+`
+	query, err := st.Prepare(queryUnit, applicationID{}, unitNameAndUUID{})
```

Comment:
**nit:** might be good to have specific custom types for these values since we do that elsewhere.

This would be a great case for SQLair to select directly into a string with a named type. We would literally be able to do `SELECT application_uuid AS ID` and pass in an `application.ID` to fill. Still want to get that ability added at some point.